File: typescript/src/events.ts
/**
 * Frames other than transactions: slot boundaries, entries and equivocation reports.
 *
 * Each is a fixed-size little-endian record with no variable tail, so reading one is a bounds check
 * and a handful of loads.
 *
 * # Why entries carry no transaction bytes
 *
 * An entry's transactions already travel on the transaction stream, each carrying the `entryIndex`
 * that puts it back in its entry. Repeating the bytes here would double the bandwidth of a
 * subscriber taking both streams to say nothing new. What the entry frame adds is the structure
 * around them: the PoH hash, the hash count, and how many transactions the entry claimed — which is
 * what lets a subscriber notice one is missing.
 */

import { ProtocolError, Verification } from './protocol.js';

function require_(bytes: Buffer, need: number, what: string): void {
  if (bytes.length < need) {
    throw new ProtocolError(`${what} needs ${need} bytes, have ${bytes.length}`);
  }
}

function decodeVerification(byte: number): Verification {
  switch (byte) {
    case 0:
      return Verification.Verified;
    case 2:
      return Verification.UnknownLeader;
    case 3:
      return Verification.StaleSchedule;
    default:
      return Verification.Disabled;
  }
}

/** Bytes in an encoded slot start. */
export const SLOT_START_LEN = 64;
/** Bytes in an encoded slot end. */
export const SLOT_END_LEN = 48;
/** Bytes in an encoded entry. */
export const ENTRY_LEN = 88;
/** Bytes in an encoded duplicate report. */
export const DUPLICATE_LEN = 160;

/** The first shred of a slot arrived. */
export interface SlotStart {
  slot: bigint;
  /** Present once a shred declaring the parent has arrived. */
  parentSlot: bigint | null;
  /** The slot's leader, when the schedule knew one. */
  leader: Buffer | null;
  /** When the first shred arrived, on the server's monotonic clock. */
  rxTsNs: bigint;
  source: number;
  shredVersion: number;
}

/** Reads a slot start. */
export function readSlotStart(bytes: Buffer): SlotStart {
  require_(bytes, SLOT_START_LEN, 'slot start');
  const flags = bytes.readUInt8(60);
  return {
    slot: bytes.readBigUInt64LE(0),
    parentSlot: (flags & 1) !== 0 ? bytes.readBigUInt64LE(8) : null,
    leader: (flags & 2) !== 0 ? bytes.subarray(16, 48) : null,
    rxTsNs: bytes.readBigUInt64LE(48),
    source: bytes.readUInt16LE(56),
    shredVersion: bytes.readUInt16LE(58),
  };
}

/** A slot finished, either completed or given up on. */
export interface SlotEnd {
  slot: bigint;
  parentSlot: bigint;
  dataShreds: number;
  fecSets: number;
  /** Sets that had to be reconstructed from parity. */
  recoveredSets: number;
  /** Shreds never seen and never recoverable. */
  missingShreds: number;
  /** When the slot was retired, on the server's monotonic clock. */
  endTsNs: bigint;
  /** Whether the slot was seen whole. False means shreds were lost. */
  complete: boolean;
}

/** Reads a slot end. */
export function readSlotEnd(bytes: Buffer): SlotEnd {
  require_(bytes, SLOT_END_LEN, 'slot end');
  return {
    slot: bytes.readBigUInt64LE(0),
    parentSlot: bytes.readBigUInt64LE(8),
    dataShreds: bytes.readUInt32LE(16),
    fecSets: bytes.readUInt32LE(20),
    recoveredSets: bytes.readUInt32LE(24),
    missingShreds: bytes.readUInt32LE(28),
    endTsNs: bytes.readBigUInt64LE(32),
    complete: (bytes.readUInt8(40) & 1) !== 0,
  };
}

/** One proof-of-history entry. */
export interface Entry {
  slot: bigint;
  parentSlot: bigint;
  rxTsNs: bigint;
  /** PoH hashes since the previous entry. Zero marks a tick-free entry. */
  numHashes: bigint;
  /** How many transactions the entry claimed — compare against what arrived. */
  txCount: bigint;
  /** The entry's PoH hash. */
  hash: Buffer;
  fecSetIndex: number;
  entryIndex: number;
  verification: Verification;
}

/** Reads an entry. */
export function readEntry(bytes: Buffer): Entry {
  require_(bytes, ENTRY_LEN, 'entry');
  return {
    slot: bytes.readBigUInt64LE(0),
    parentSlot: bytes.readBigUInt64LE(8),
    rxTsNs: bytes.readBigUInt64LE(16),
    numHashes: bytes.readBigUInt64LE(24),
    txCount: bytes.readBigUInt64LE(32),
    hash: bytes.subarray(40, 72),
    fecSetIndex: bytes.readUInt32LE(72),
    entryIndex: bytes.readUInt32LE(76),
    verification: decodeVerification(bytes.readUInt8(80)),
  };
}

/**
 * Two different shreds arrived for one slot and index — the leader equivocated.
 *
 * Both signatures are carried so the report can be checked independently.
 */
export interface Duplicate {
  slot: bigint;
  index: number;
  firstSource: number;
  secondSource: number;
  /** When the conflict was noticed, on the server's monotonic clock. */
  detectedNs: bigint;
  /** Whether the conflicting shreds were data shreds rather than parity. */
  isData: boolean;
  firstSignature: Buffer;
  secondSignature: Buffer;
}

/** Reads a duplicate report. */
export function readDuplicate(bytes: Buffer): Duplicate {
  require_(bytes, DUPLICATE_LEN, 'duplicate');
  return {
    slot: bytes.readBigUInt64LE(0),
    index: bytes.readUInt32LE(8),
    firstSource: bytes.readUInt16LE(12),
    secondSource: bytes.readUInt16LE(14),
    detectedNs: bytes.readBigUInt64LE(16),
    isData: bytes.readUInt8(24) !== 0,
    firstSignature: bytes.subarray(32, 96),
    secondSignature: bytes.subarray(96, 160),
  };
}

/**
 * A shred republished exactly as it arrived.
 *
 * # What this stream is
 *
 * Every other event is a *conclusion* the node reached — a transaction it decoded, a slot it judged
 * finished. This is the evidence those were drawn from, handed to you at the moment the node
 * received it, before it verified or decoded anything.
 *
 * **Nothing here is claimed to be leader-signed.** A shred is republished as soon as it survives
 * deduplication; the signature check happens later, on the FEC set as a whole. If you need the
 * guarantee, take `transactions`, which carries it. If you need the latency and run your own
 * pipeline, this is for you.
 *
 * Two things it deliberately is not. **Not recovered**: a shred rebuilt from parity is not
 * republished, so a gap here is a real loss rather than an artefact, and your own count matches the
 * network's. **Not filtered**: a shred is a fragment of an erasure set, not yet anything a filter
 * can match on.
 *
 * It is the highest-rate stream a node publishes. Take it only if you are consuming it.
 */
export interface RawShred {
  /** Slot the shred belongs to. */
  slot: bigint;
  /**
   * Index within the slot, scoped by `isData`.
   *
   * A data shred and a coding shred may share an index. Key on the pair, not on the index alone.
   */
  index: number;
  /** Index of the first data shred of the FEC set this belongs to. */
  fecSetIndex: number;
  /** When the node received it, on the server's monotonic clock. */
  rxTsNs: bigint;
  /** Which configured source delivered it first. */
  source: number;
  /** Whether this is a data shred rather than parity. */
  isData: boolean;
  /** The shred exactly as it arrived, ready for a parser that expects one. */
  bytes: Buffer;
}

/** Bytes before a raw shred's payload. */
export const RAW_SHRED_HEADER_LEN = 32;

/** Decodes a republished shred, borrowing the payload rather than copying it. */
export function readRawShred(bytes: Buffer): RawShred {
  require_(bytes, RAW_SHRED_HEADER_LEN, 'raw shred');
  return {
    slot: bytes.readBigUInt64LE(0),
    index: bytes.readUInt32LE(8),
    fecSetIndex: bytes.readUInt32LE(12),
    rxTsNs: bytes.readBigUInt64LE(16),
    source: bytes.readUInt16LE(24),
    // Anything other than the parity marker is data: a kind byte this build does not know is a
    // newer node, and refusing the frame would break you over a field you do not use.
    isData: bytes.readUInt8(26) !== 1,
    bytes: bytes.subarray(RAW_SHRED_HEADER_LEN),
  };
}

File: rust/src/events.rs
//! Frames other than transactions: slot boundaries, entries and equivocation reports.
//!
//! Each is a fixed-size little-endian record with no variable tail, so reading one is a bounds
//! check and a handful of loads.
//!
//! # Why entries carry no transaction bytes
//!
//! An entry's transactions already travel on the transaction stream, each carrying the
//! `entry_index` that puts it back in its entry. Repeating the bytes here would double the
//! bandwidth of a subscriber taking both streams to say nothing new. What the entry frame adds is
//! the structure around them: the PoH hash, the hash count, and how many transactions the entry
//! claimed — which is what lets a subscriber notice one is missing.

use crate::{
    error::{Error, Result},
    protocol::Verification,
};

fn require(bytes: &[u8], need: usize, what: &'static str) -> Result<()> {
    if bytes.len() < need {
        return Err(Error::Truncated {
            what,
            need,
            have: bytes.len(),
        });
    }
    Ok(())
}

fn u16_at(bytes: &[u8], at: usize) -> u16 {
    u16::from_le_bytes([bytes[at], bytes[at + 1]])
}

fn u32_at(bytes: &[u8], at: usize) -> u32 {
    u32::from_le_bytes(bytes[at..at + 4].try_into().expect("bounds checked"))
}

fn u64_at(bytes: &[u8], at: usize) -> u64 {
    u64::from_le_bytes(bytes[at..at + 8].try_into().expect("bounds checked"))
}

/// The first shred of a slot arrived.
#[derive(Clone, Copy, Debug)]
pub struct SlotStart<'a> {
    /// Slot that just appeared.
    pub slot: u64,
    /// Present once a shred declaring the parent has arrived.
    pub parent_slot: Option<u64>,
    /// The slot's leader, when the schedule knew one.
    pub leader: Option<&'a [u8; 32]>,
    /// When the first shred arrived, on the server's monotonic clock.
    pub rx_ts_ns: u64,
    /// Which ingress source delivered it.
    pub source: u16,
    /// Shred version carried by the first shred.
    pub shred_version: u16,
}

impl<'a> SlotStart<'a> {
    /// Bytes in the encoded frame.
    pub const LEN: usize = 64;

    /// Reads a slot start.
    pub fn read(bytes: &'a [u8]) -> Result<Self> {
        require(bytes, Self::LEN, "slot start")?;
        let flags = bytes[60];
        Ok(Self {
            slot: u64_at(bytes, 0),
            parent_slot: (flags & 1 != 0).then(|| u64_at(bytes, 8)),
            leader: (flags & 2 != 0).then(|| bytes[16..48].try_into().expect("bounds checked")),
            rx_ts_ns: u64_at(bytes, 48),
            source: u16_at(bytes, 56),
            shred_version: u16_at(bytes, 58),
        })
    }
}

/// A slot finished, either completed or given up on.
#[derive(Clone, Copy, Debug)]
pub struct SlotEnd {
    /// Slot that finished.
    pub slot: u64,
    /// Parent of `slot`.
    pub parent_slot: u64,
    /// Data shreds observed or recovered.
    pub data_shreds: u32,
    /// FEC sets that completed.
    pub fec_sets: u32,
    /// Sets that had to be reconstructed from parity.
    pub recovered_sets: u32,
    /// Shreds never obtained, so their entries were never emitted.
    pub missing_shreds: u32,
    /// When the slot was retired, on the server's monotonic clock.
    pub end_ts_ns: u64,
    /// Whether the slot was seen whole. False means shreds were lost.
    pub complete: bool,
}

impl SlotEnd {
    /// Bytes in the encoded frame.
    pub const LEN: usize = 48;

    /// Reads a slot end.
    pub fn read(bytes: &[u8]) -> Result<Self> {
        require(bytes, Self::LEN, "slot end")?;
        Ok(Self {
            slot: u64_at(bytes, 0),
            parent_slot: u64_at(bytes, 8),
            data_shreds: u32_at(bytes, 16),
            fec_sets: u32_at(bytes, 20),
            recovered_sets: u32_at(bytes, 24),
            missing_shreds: u32_at(bytes, 28),
            end_ts_ns: u64_at(bytes, 32),
            complete: bytes[40] & 1 != 0,
        })
    }
}

/// One proof-of-history entry.
#[derive(Clone, Copy, Debug)]
pub struct Entry<'a> {
    /// Slot the entry was produced in.
    pub slot: u64,
    /// Parent of `slot`.
    pub parent_slot: u64,
    /// When the FEC set's first shred arrived, on the server's monotonic clock.
    pub rx_ts_ns: u64,
    /// PoH hashes since the previous entry. Zero means this is not a tick boundary.
    pub num_hashes: u64,
    /// Transactions the entry claims — compare against what arrived.
    pub tx_count: u64,
    /// The entry's PoH hash.
    pub hash: &'a [u8; 32],
    /// FEC set the entry came from.
    pub fec_set_index: u32,
    /// Index of the entry within its batch.
    pub entry_index: u32,
    /// Signature status of the FEC set.
    pub verification: Verification,
}

impl<'a> Entry<'a> {
    /// Bytes in the encoded frame.
    pub const LEN: usize = 88;

    /// Reads an entry.
    pub fn read(bytes: &'a [u8]) -> Result<Self> {
        require(bytes, Self::LEN, "entry")?;
        Ok(Self {
            slot: u64_at(bytes, 0),
            parent_slot: u64_at(bytes, 8),
            rx_ts_ns: u64_at(bytes, 16),
            num_hashes: u64_at(bytes, 24),
            tx_count: u64_at(bytes, 32),
            hash: bytes[40..72].try_into().expect("bounds checked"),
            fec_set_index: u32_at(bytes, 72),
            entry_index: u32_at(bytes, 76),
            verification: Verification::from_byte(bytes[80]),
        })
    }
}

/// Two different shreds arrived for one slot and index — the leader equivocated.
///
/// Both signatures are carried so the report can be checked independently; which fork survives is
/// not knowable at this layer, and a subscriber may care about the conflict itself.
#[derive(Clone, Copy, Debug)]
pub struct Duplicate<'a> {
    /// Slot both shreds claim.
    pub slot: u64,
    /// Index both shreds claim.
    pub index: u32,
    /// Source that delivered the version that arrived first.
    pub first_source: u16,
    /// Source that delivered the conflicting version.
    pub second_source: u16,
    /// When the conflict was noticed, on the server's monotonic clock.
    pub detected_ns: u64,
    /// Whether the conflicting shreds were data shreds rather than parity.
    pub is_data: bool,
    /// Signature of the version that arrived first.
    pub first_signature: &'a [u8; 64],
    /// Signature of the conflicting version.
    pub second_signature: &'a [u8; 64],
}

impl<'a> Duplicate<'a> {
    /// Bytes in the encoded frame.
    pub const LEN: usize = 160;

    /// Reads a duplicate report.
    pub fn read(bytes: &'a [u8]) -> Result<Self> {
        require(bytes, Self::LEN, "duplicate")?;
        Ok(Self {
            slot: u64_at(bytes, 0),
            index: u32_at(bytes, 8),
            first_source: u16_at(bytes, 12),
            second_source: u16_at(bytes, 14),
            detected_ns: u64_at(bytes, 16),
            is_data: bytes[24] != 0,
            first_signature: bytes[32..96].try_into().expect("bounds checked"),
            second_signature: bytes[96..160].try_into().expect("bounds checked"),
        })
    }
}

/// A shred republished exactly as it arrived.
///
/// # What this stream is
///
/// Every other event is a *conclusion* the node reached — a transaction it decoded, a slot it
/// judged finished. This is the evidence those were drawn from, handed to you at the moment the
/// node received it, before it verified or decoded anything.
///
/// **Nothing here is claimed to be leader-signed.** A shred is republished as soon as it survives
/// deduplication; the signature check happens later, on the FEC set as a whole. If you need the
/// guarantee, take `transactions`, which carries it. If you need the latency and run your own
/// pipeline, this is for you.
///
/// Two things it deliberately is not. **Not recovered**: a shred rebuilt from parity is not
/// republished, so a gap here is a real loss rather than an artefact, and your own count matches the
/// network's. **Not filtered**: a shred is a fragment of an erasure set, not yet anything a filter
/// can match on.
///
/// It is the highest-rate stream a node publishes — a multiple of `transactions`, since most
/// fragments carry votes. Take it only if you are consuming it.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct RawShred<'a> {
    /// Slot the shred belongs to.
    pub slot: u64,
    /// Index within the slot, scoped by [`Self::is_data`].
    ///
    /// A data shred and a coding shred may share an index. Key on the pair, not on the index alone.
    pub index: u32,
    /// Index of the first data shred of the FEC set this belongs to.
    pub fec_set_index: u32,
    /// When the node received it, on the server's monotonic clock.
    pub rx_ts_ns: u64,
    /// Which configured source delivered it first.
    pub source: u16,
    /// Whether this is a data shred rather than parity.
    pub is_data: bool,
    /// The shred exactly as it arrived, ready for a parser that expects one.
    pub bytes: &'a [u8],
}

impl<'a> RawShred<'a> {
    /// Bytes before the payload.
    pub const HEADER_LEN: usize = 32;

    /// Reads the frame, borrowing the payload rather than copying it.
    pub fn read(bytes: &'a [u8]) -> Result<Self> {
        if bytes.len() < Self::HEADER_LEN {
            return Err(Error::Truncated {
                what: "raw shred",
                need: Self::HEADER_LEN,
                have: bytes.len(),
            });
        }
        Ok(Self {
            slot: u64::from_le_bytes(bytes[0..8].try_into().expect("in range")),
            index: u32::from_le_bytes(bytes[8..12].try_into().expect("in range")),
            fec_set_index: u32::from_le_bytes(bytes[12..16].try_into().expect("in range")),
            rx_ts_ns: u64::from_le_bytes(bytes[16..24].try_into().expect("in range")),
            source: u16::from_le_bytes(bytes[24..26].try_into().expect("in range")),
            // Anything other than the parity marker is data: a kind byte this build does not know
            // is a newer node, and refusing the frame would break you over a field you do not use.
            is_data: bytes[26] != 1,
            bytes: &bytes[Self::HEADER_LEN..],
        })
    }
}
